src/linalg.h

```diff
@@ -75,7 +80,27 @@ inline void matrix_multiplication(T** matrix, T* vector_x,
 
 
 /**
- * @brief Scale vectgor by some scalar value.
+ * @brief Transpose a 2D matrix.
+ * @param matrix array to transpose
+ * @param dim1 first dimension length
+ * @param dim2 second dimension length
+ */
+template<typename T>
+inline void matrix_transpose(T* matrix, int dim1, int dim2) {
+
+  std::vector<T> temp;
+
+  for (int i=0; i < dim1; i++) {
+    for (int j=0; j < dim2; j++)
+      temp.push_back(matrix[j * dim1 + i]);
+  }
```

Comment:
It might not make a huge difference, but this is an inefficient way to make a transpose. We know the size of the matrix we need to make so we can allocate all the memory upfront. Also I believe the linear algebra folks would tell you to reference the same data - just with a different indexer to remove unnecessary extra storage of data.